packages/matomo/index.tsx

```diff
@@ -1,13 +1,110 @@
+import { CartLineItem } from '@ifixit/cart-sdk';
+import { Money, sumMoney } from '@ifixit/helpers';
+
+const _paq: any[] | undefined =
+   typeof window === 'undefined' ? undefined : (window as any)._paq;
+
 /**
  * @see https://developer.matomo.org/api-reference/tracking-javascript
  */
+export function trackPageView(url: string) {
+   if (!_paq) {
+      return;
+   }
+   _paq.push(['setCustomUrl', url]);
+   _paq.push(['setDocumentTitle', document.title]);
+   _paq.push(['trackPageView']);
+}
 
-declare const _paq: string[][] | undefined;
+/**
+ * @see https://developer.matomo.org/api-reference/tracking-javascript
+ * @see https://matomo.org/docs/ecommerce-analytics/#tracking-product-views-in-matomo
+ */
+type ProductData = {
+   productSku: string;
+   productName?: string;
+   /**
+    * Category name, or up to five unique categories, e.g. ["Books", "New
+    * Releases", "Technology"]
+    */
+   categoryName?: string | [string, string?, string?, string?, string?];
+   price: Money;
+};
 
-export function trackPageView(url: string) {
-   if (typeof window !== 'undefined' && _paq) {
-      _paq.push(['setCustomUrl', url]);
-      _paq.push(['setDocumentTitle', document.title]);
-      _paq.push(['trackPageView']);
+export function trackMatomoEcommerceView(product: ProductData) {
+   if (!_paq) {
+      return;
+   }
+   _paq.push([
+      'setEcommerceView',
+      product.productSku,
+      product.productName,
+      product.categoryName,
+      product.price.amount,
+   ]);
+}
+
+export function trackMatomoCartChange(items: CartLineItem[]) {
+   if (!_paq) {
+      return;
+   }
+   trackClearCart();
```

Comment:
This seems wrong. Why do we clear and re-add every time the cart is changed? Seems like that's going to add a lot of fake events to Matomo.

Reply:
This is how we do it in ifixit https://github.com/iFixit/ifixit/blob/master/carpenter-frontend/Shared/Analytics/Matomo/subcribeToMiniCart.ts.

@jarstelfox can you remind us why it's done this way in iFixit/ifixit?

Reply:
Matomo is stateful: every time you call out to Matomo saying you are adding an item, it will add the state it it’s view.

The problem on the main site was keeping cart data in Matomo completely in sync with the cart state. Mainly, the hook I was using to grab the cart state just tells me a user has 1 of X and 1 of Y in their cart, not that the user only now just added Y to the cart.

Full example:

Cart empty. User views page X and adds X to cart. Fire cart={ X: 1 } to Matomo. Matomo’s cart: { X: 1 } User views page Y and adds Y to cart. Fire cart: { X: 1, Y: 1}. Matomo’s cart: { X: 2, Y: 1 }

In fact, the problem with the main repo’s code is the cart update event was fired on every page load since we grab that data over the wire.

In short, this could all be fixed with some state diffing, but this was the safest way to shop for V1

Reply:
Thanks for the explanation! Since we're ok with this behavior on the current product page, I'm inclined to say that it's fine to keep the status quo for V1 of the Next.js product page. We can devote more time to this after we launch.